dsgrid/config/annual_time_dimension_config.py

```diff
@@ -161,15 +112,45 @@ def get_tzinfo(self):
     def get_time_interval_type(self):
         return None
 
-    def list_expected_dataset_timestamps(self, model_years=None, data_adjustment=None):
-        if model_years is not None:
-            # We do not expect to need this.
-            raise NotImplementedError(
-                f"No support for {model_years=} in {type(self)}.list_expected_dataset_timestamps"
-            )
-
+    def list_expected_dataset_timestamps(self, data_adjustment=None):
         timestamps = []
         for time_range in self.model.ranges:
             start, end = (int(time_range.start), int(time_range.end))
             timestamps += [AnnualTimestampType(x) for x in range(start, end + 1)]
         return timestamps
+
+
+def map_annual_historical_time_to_date_time(
+    df: DataFrame,
+    annual_dim: AnnualTimeDimensionConfig,
+    dt_dim: DateTimeDimensionConfig,
+    value_columns: set[str],
+) -> DataFrame:
+    """Map a DataFrame with an annual/historical time dimension to a DateTime time dimension."""
+    annual_col = annual_dim.get_load_data_time_columns()[0]
+    dt_df = dt_dim.build_time_dataframe()
+    myear_column = DimensionType.MODEL_YEAR.value
+    dt_col = dt_dim.get_load_data_time_columns()[0]
+
+    # Note that MeasurementType.TOTAL has already been verified.
+    with custom_spark_conf(
+        {"spark.sql.session.timeZone": dt_dim.model.datetime_format.timezone.tz_name}
+    ):
+        years = dt_df.withColumn("year", F.year(dt_col)).select("year").distinct().collect()
+        if len(years) != 1:
```

Comment:
Is this acceptable? I'm not sure what to do if the project time spans multiple years.

Reply:
Yes, acceptable.